source/nodejs/adaptivecards-designer/src/designer-peers.ts

```diff
@@ -962,6 +962,123 @@ class NameValuePairPropertyEditor extends PropertySheetEntry {
     }
 }
 
+class InnerStructPropertyEditor extends PropertySheetEntry {
+    private collectionChanged(context: PropertySheetContext, innerPropertiesList: string[][], refreshPropertySheet: boolean) {
+        context.target[this.collectionPropertyName] = [];
+
+        for (let innerProperties of innerPropertiesList) {
+            let item = this.createCollectionItem(innerProperties);
+
+            context.target[this.collectionPropertyName].push(item);
+        }
+
+        context.peer.changed(refreshPropertySheet);
+    }
+
+    render(context: PropertySheetContext): Adaptive.CardElement {
+        let result = new Adaptive.Container();
+
+        let collection = context.target[this.collectionPropertyName];
+
+        if (!Array.isArray(collection)) {
+            throw new Error("The " + this.collectionPropertyName + " property on " + context.peer.getCardObject().getJsonTypeName() + " either doesn't exist or isn't an array.")
+        }
+
+        let innerPropertiesList: string[][] = [];
+
+        for (let innerProperties of collection) {
+            var newItem : string[] = [];
+            for (let i = 0; i < this.innerPropertiesLength; i++)
+            {
+                newItem.push(innerProperties[this.innerPropertiesNames[i]]);
+            }
+            innerPropertiesList.push(newItem);
+        }
+
+        if (innerPropertiesList.length == 0) {
+            let messageTextBlock = new Adaptive.TextBlock();
+            messageTextBlock.spacing = Adaptive.Spacing.Small;
+            messageTextBlock.text = this.messageIfEmpty;
+
+            result.addItem(messageTextBlock);
+        }
+        else {
+            for (let i = 0; i < innerPropertiesList.length; i++) {
+                let columnSet = new Adaptive.ColumnSet();
+                columnSet.spacing = Adaptive.Spacing.Small;
+
+                for (let j = 0; j < this.innerPropertiesLength; j++) {
```

Comment:
Problem here is that you're going to have three (or more) input on one single line, that doesn't scale given the restricted space available in the "Element Properties" pane. We need to have logic that switches to a better layout when there are more than two inputs, something like the below:
```json
{
    "type": "ColumnSet",
    "separator": true,
    "columns": [
        {
            "type": "Column",
            "width": "stretch",
            "items": [
                {
                    "type": "Input.Text",
                    "placeholder": "Label for first input"
                },
                {
                    "type": "Input.Text",
                    "placeholder": "Label for second input"
                },
                {
                    "type": "Input.Text",
                    "placeholder": "Label for thrid input"
                }
            ]
        },
        {
            "type": "Column",
            "width": "auto",
            "items": [
                {
                    "type": "ActionSet",
                    "actions": [
                        {
                            "type": "Action.OpenUrl",
                            "title": "X"
                        }
                    ]
                }
            ],
            "verticalContentAlignment": "Center"
        }
    ]
}
```

which renders like this:
![image](https://user-images.githubusercontent.com/1334689/161865009-596d7393-9fbb-45e6-a118-275a2d029873.png)

Reply:
I've changed it to always render vertically (screenshot of what it looks like below). That way we don't need to render differently based on the number of properties. 

<img width="292" alt="sources" src="https://user-images.githubusercontent.com/26824113/161868051-13075111-96df-4753-81c4-4d3826c0ea41.png">

Reply:
I suppose that's fine. Thanks for adding designer support!